src/main/scala/org/kframework/minikore/interfaces/MiniKoreInterface.scala

```diff
@@ -0,0 +1,303 @@
+package org.kframework.minikore.interfaces
+
+object TreeInterface {
+
+  import PatternInterface._
+
+  sealed trait AST
+
+
+  sealed trait Node extends AST with Product {
+    def args: Seq[Pattern]
+
+    def build(children: Seq[Pattern]): Pattern
+  }
+
+  object Node {
+    def unapply(arg: Node): Option[Seq[Pattern]] = Some(arg.args)
+  }
+
+
+  sealed trait Leaf[C] extends AST with Product {
+    def contents: C
+
+    def build(contents: C): Pattern
+  }
+
+  object Leaf {
+    def unapply[C](arg: Leaf[C]): Option[C] = Some(arg.contents)
+  }
+
+
+  sealed trait Leaf2[CC1, CC2] extends Leaf[Product2[CC1, CC2]] with Product2[CC1, CC2] {
+    override def contents: Product2[CC1, CC2] = (_1, _2)
+
+    def apply(_1: CC1, _2: CC2): Pattern
```

Comment:
Yes, but that apply method is added on the companion, as it should, not on the instances of the class.

```
case class Foo(x: Int)
```

generates:

```
object Foo {
  def apply(x: Int) = new Foo(x)
}

class Foo(val x: Int) {
  def equals ...
  def hashcode ...
}
```

The class `Foo` does not have an `apply` method, the companion object `Foo` does. So `Foo(1)` works by calling the `apply` on the companion object, while `new Foo(1)(2)` because the instance itself does not have an `apply` method.